### ISSUE TITLE
Request for additional preference types

### ISSUE THREAD

**Issue by Gericop** [Gericop/Android-Support-Preference-V7-Fix#87]
New preference types are added occasionally to a separate artifact:
```gradle
compile 'com.takisoft.fix:preference-v7-extras:+'
```
#### Current and planned extra preference types:
- [x] RingtonePreference (available since `26.0.1.1-alpha2`)
- [ ] TimePickerPreference
- [ ] DatePickerPreference

---

### If you have any requests about new preference types, add them here

**Comment by DreaminginCodeZH** [comment]
SeekBarPreference: You can base it upon [my impl](https://github.com/DreaminginCodeZH/SeekBarPreference).

**Comment by Gericop** [comment]
@DreaminginCodeZH `SeekBarPreference` is already available both in the official lib and the fixed lib as well.

**Comment by DreaminginCodeZH** [comment]
Maybe a `SimpleMenuPreference` at [here](https://github.com/RikkaW/MaterialPreference/blob/master/preference-simplemenu/src/main/java/moe/shizuku/preference/SimpleMenuPreference.java)? It correctly implemented the Material Design spec, but is based on the framework Preference and compilation hacks.

**Comment by Gericop** [comment]
Isn't it the `DropDownPreference`? Because it looks like that.

**Comment by DreaminginCodeZH** [comment]
@Gericop 

No:

- Its padding is correct (and should size on the steps).
- It shows the selected item as selected.
- It positions the menu correctly, making the selected item appear above the preference.

[Simple Menu - Material Design Guidelines](https://material.io/guidelines/components/menus.html#menus-menu-items)

**Comment by Gericop** [comment]
Well, the DatePickerPreference and TimePickerPreference are almost done, I had to port the entire framework version of the picker dialogs to show material themed pickers (I know about the separate [datetimepicker project](https://android.googlesource.com/platform/frameworks/opt/datetimepicker/) of AOSP, but that uses the first (=old) version of the material pickers, and had to solve some problems related to date and time formatting on older APIs (mainly 17 and under). It's only in a [Github repo](https://github.com/Gericop/DateTimePicker) for the time being, but soon it's gonna be released to jCenter, too.

**Comment by Gericop** [comment]
The 3 new preference types are now available on jCenter.
> The `preference-v7-extras` artifact is not available yet, but it should be in a couple of hours.

[`RingtonePreference`](https://github.com/Gericop/Android-Support-Preference-V7-Fix/wiki/Preference-types#ringtonepreference):
```gradle
compile 'com.takisoft.fix:preference-v7-ringtone:26.0.2.0'
```

[`DatePickerPreference`](https://github.com/Gericop/Android-Support-Preference-V7-Fix/wiki/Preference-types#datepickerpreference) and [`TimePickerPreference`](https://github.com/Gericop/Android-Support-Preference-V7-Fix/wiki/Preference-types#timepickerpreference):
```gradle
compile 'com.takisoft.fix:preference-v7-datetimepicker:26.0.2.0'
```

**Comment by DreaminginCodeZH** [comment]
@Gericop Is the `SimpleMenuPreference` going to land in this library? If not, I'm also considering creating a separate library for it.

**Comment by Gericop** [comment]
@DreaminginCodeZH I'll look into it. I'm finishing the `ColorPickerPreference` now, then check out what's needed for the simple menus.

**Comment by Gericop** [comment]
The funny thing is that `DropDownPreference` has everything basically to implement this (so Google could have / should have done it), but the most important part (the item measuring) is package private, so it means that either a _hack_ is needed (placing the custom preference to a package that has the same name as the official one), or the whole class must be duplicated to override that functionality. Seriously, sometimes I just don't understand why would they choose to make some methods like this (package) private...

**Comment by opus1269** [comment]
Would it be possible to support any custom preference that extends from `DialogPreference`?

I have a need for a `NumberPicker ` preference. I could do it with a list, but the `NumberPicker` is cleaner IMO.

**Comment by Gericop** [comment]
@opus1269 Yeah, it is currently undocumented, but you can add custom `DialogPreference` implementations by extending `DialogPreference` for the preference and `PreferenceDialogFragmentCompat` for the dialog, and adding the class references using this:

```java
public class YourCustomPreference extends DialogPreference {
    static {
        PreferenceFragmentCompat.addDialogPreference(YourCustomPreference.class, YourCustomPreferenceDialogFragmentCompat.class);
    }
    // ...
}
```
The lib's `PreferenceFragmentCompat` implementation will handle and display the given dialog fragment if the preference is touched by the user.

`NumberPickerPreference` is certainly not a bad idea, but you need a material picker because the internal widget does not comply with the MD guidelines on pre-Lollipop devices. The [DateTimePicker](https://github.com/Gericop/DateTimePicker) lib contains an implementation (the official one from the Android 8.0 source) since the collapsed view shows a spinner (a `NumberPicker`) instead of a calendar view, however, it is currently not available externally.

**Comment by DreaminginCodeZH** [comment]
Can you make the `addDialogPreference` method accept any `Fragment`? It would make things easier for displaying an `AppCompatDialogFragment` without the need for a key on the preference (or else the support lib tries to `findPreference` which will throw exception in `onCreate` of `PreferenceDialogFragmentCompat`).

**Comment by Gericop** [comment]
Is there any particular reason why you don't want to extend `PreferenceDialogFragmentCompat` instead of `AppCompatDialogFragment`? `PreferenceDialogFragmentCompat` is meant to be used for `DialogPreference`'s since it provides some helpful things such as `getPreference()` which can be used to query and update the preference itself.

**Comment by DreaminginCodeZH** [comment]
> without the need for a key on the preference (or else the support lib tries to findPreference which will throw exception in onCreate of PreferenceDialogFragmentCompat).

**Comment by Gericop** [comment]
I know that but why wouldn't you want a key for a preference? What's the use case?

**Comment by DreaminginCodeZH** [comment]
Please see edit on my last comment.

**Comment by Gericop** [comment]
I see. Well, it can be done for sure. I'll probably release a new version later today with this, the colliding attribute names fixed, and some other fixes.

**Comment by Gericop** [comment]
Updating now the README.md so it can be released.

**Comment by Gericop** [comment]
Released v**26.1.0.3**.

**Comment by DreaminginCodeZH** [comment]
Hi, I just checked out 26.1.0.3 and found that the second argument is still requiring a `DialogFragment`, but in fact one can choose to use a fragment to receive `onActivityResult()`. This technique can be used in for example a `RingtonePreference` that launches the stock ringtone chooser activity. So can you just relax the constraint to be a `Fragment` to make it more flexible?

**Comment by Gericop** [comment]
That wouldn't make sense semantically since a `DialogPreference` should display a dialog (I know that the ringtone picker Activity looks like a dialog, but it's an external component to the app). 
Also, there's no way of knowing when the given `Fragment` should be removed from the `FragmentManager` meaning the developer should handle it which can lead to bugs, so for both of these reasons I will not relax the constraint.
_However_, you can do a little hack and use the `DialogFragment` just like if it was a normal `Fragment`:

1. Create the fragment where you override `onCreateDialog(...)` so that it returns `null` (this is allowed by `DialogFragment`)
2. Launch whatever `Activity` you want from your fragment
3. In `onActivityResult(...)` after processing the results, call `dismiss()`.

This will add the `DialogFragment` to the `FragmentManager` listening to whatever result you expect, then the `dismiss()` call removes it from the `FragmentManager`.

**Comment by DreaminginCodeZH** [comment]
I think API design should not prevent developer from accomplish certain task even if it introduces likelyhood of error - error-prone developers won't try to understand why the limitation is there and will work around it in some other ugly way, while people with true need of an API will be forced to hack. Intead, this should be the work of documentation.

As for semantics - My `RingtonePreference` isn't a `DialogPreference`, and the support library did not require a `DialogPreference` in `onDisplayPreferenceDialog()`. Simply any `Preference` that want to display a "dialog" can do that, and there's no need to prevent people from doing what they see fit in their own situation.

Or at least, please consider adding a `@Deprecated` overload that takes a `Fragment` and explain the reason there  - it's painful to add another wrapper over `PreferenceFragmentCompat` and duplicate the logic in every app just because of a type constraint that could have been relaxed.

**Comment by Gericop** [comment]
You can do what I suggested in the 3-steps thingy. It's essentially the same what you want but prevents people from _accidentally_ adding any kind of `Fragment` without understanding the problem it introduces. Still, this whole "use a proxy `Fragment` to open an `Activity` to get some results back to the preference" seems odd to me. I'd rather look for a better solution.

I chose not to use the ringtone picker `Activity` because on pre-Lollipop it doesn't have the material theme, on Lollipop and up it has the material theme but the color scheme doesn't match the app's one, so the UX is bad, and also it doesn't have things like "add ringtone" on older APIs.

**Comment by DreaminginCodeZH** [comment]
`onCreateDialog()` is annotated with `@NonNull` in support-v4 `DialogFragment`, although current behavior guarded against null, I don't think it's guaranteed and prefer not doing so. So the 3-steps thing is not a correct workaround for this, not to mention it is far too much incorrect both semantically and in implementation.

Using a headless `Fragment` is not an odd thing, but [a practice](https://www.google.com/search?q=android+headless+fragment) that has been around for a while. In contrast `onCreateView()` is marked with `@Nullable`, and it is a common technique to use a `Fragment` without UI to retain state of network call during configuration changes.

And just to stay consistent with support library, it didn't put restriction on the developer to extend `DialogPreference` to show any form of a dialog (see PreferenceFragmentCompat#onDisplayPreferenceDialog()). You cannot prevent developers from overriding their own `onDisplayPreferenceDialog()`, so they can always do insane things. And this is also a hidden API which dumb developers won't find. Why don't just be a bit more open about API and believe developers? After all, this is only a (super-)convenient wrapper around `onDisplayPreferenceDialog(Preference preference)`.

So can you be so kind to at least add a deprecated overload, so that a stupid artifact just adding one method call won't ever be needed on maven? It can be like this:

```java
    /**
     * @deprecated Most users should add a {@link DialogPreference} with a
     * {@link PreferenceDialogFragmentCompat}. If you are using this overload, make sure you remove
     * your fragment once work is done.
     */
    public static void addDialogPreference(Class<? extends Preference> preferenceClass,
                                           Class<? extends Fragment> fragmentClass) {
        ...
```

---

I acknowledge that the custom ringtone picker approach has its advantages, but its disadvantages concerns me and I'd like to have the freedom for an alternative.

**Comment by Gericop** [comment]
Technically yes, there's no restriction on `DialogPreference`, but semantically this is odd and _should_ use only `DialogPreference` because, well, you want to display a dialog. The `DialogPreference`'s `onClick()` calls the `PreferenceManager`'s `showDialog(...)` which then calls `onDisplayPreferenceDialog(...)`. The original implementation of `onDisplayPreferenceDialog(...)` expects the returned fragment to be a `DialogFragment` so you can see that the devs obviously expect a dialog to be displayed there despite the generic `Preference` approach (I know that in the overriding class I use `Fragment` but still, the originally intended use is pretty clear here).

I know headless fragments and I wouldn't call this a headless but a dangling one because it just does not belong anywhere. It's basically used for hacking the system and starting an `Activity`.

As I said, I'm not changing this right now because I feel like this is a bad approach (beside the fact that this is clearly meant to be for internal use only, at least for the time being) and will investigate what better options would be available. Be patient...

**Comment by Gericop** [comment]
Here's a possible approach:

1. The custom preference implements an interface that has two methods: one for receiving the `PreferenceFragmentCompat` instance so it can start the `Activity` itself, and one for receiving the results (same signature as `onActivityResult(...)`)
2. The fragment's `onActivityResult(...)` iterates over all of the preferences and calls `onActivityResult(...)` on all preferences until the appropriate one is found

This removes the burden from the dev having to create a separate fragment and ensures that no dangling fragments are left behind upon returning from the target `Activity`. A possible drawback could be the case when the request codes equal in two different preferences but the request codes should also be user settable since they can also use the same request code elsewhere which could lead to problems (hence there's a `setCustomRingtoneRequestCode(...)` method in the `RingtonePreference`).

**Comment by DreaminginCodeZH** [comment]
It's not a hack but a representation of a dialog that is displayed by an external activity. The second proposed approach is also what I have considered before, and I think it burdens developers by taking away one specific request code from them and forcing them to write the callback logic in every preference fragment they use the preference, so I take the `Fragment` approach.

On the contrary, using a fragment to represent the existence of a dialog activity and track its state is semantically correct, and leverages support library's implementation of `Fragment.onActivitiyResult()` which won't be delivered to any other `Fragment` and is self-contained which won't pollute developer's preference fragment code. It is not dangling because it spawns with the extern dialog activity and is removed when the activity finishes, that is, it is a representation of the existence of an external dialog.

Neither `PreferenceFragmentCompat.onDisplayPreferenceDialog()` nor `PreferenceManager.showDialog()` required a `DialogPreference` instead of a general `Preference`. So I believe when considering the design of a fixing library no extra constraints should be imposed on developers. Any `Preference` should be able to call `showDialog()` easily, and any `Fragment` (or anything) that can present a dialog to user in any form should still be allowed in a wrapper for `onDisplayPreferenceDialog`. In a word I think when designing a library it should tend to be more flexible instead of more restrictive, because there can always be some possible truly valid use cases that become prevented, and bad developers can always do bad things in other bad ways.

**Comment by Gericop** [comment]
The second approach does not take anything away. Since the preference receives a `Fragment` instance, it is more than welcome to do whatever it wants (add fragments to it for example), but it also makes it easier for those who just want to launch an `Activity` and get the results back without implementing extra fragments and whatnot. Another good thing about this approach is that the devs can control _how_ (`commit()`, `commitNow()`, etc.) they want to add the fragment (if any).

This lib is technically not restricting anything. Yes, it (still an _internal use only_ method, for the time being) requires a `DialogPreference` because it is meant to be used with those type of preferences (since they call the dialog showing logic on click by default). If I went back to the official way, the `onDisplayPreferenceDialog()` would only accept `DialogFragment` classes when instantiating them. Sure I can make the `addDialogPreference(...)` to accept a generic `Preference` just to satisfy the no-more-restriction need, but the second parameter would remain `DialogFragment` because **the intended use is to display a dialog fragment for the preference** unless the logic is changed there too.

If you still insist, sure, I can make it happen to add a `Fragment` via `addDialogPreference(...)` but I would prefer the second approach (the interface one) because it is more robust and leaves the originally intended, easy-to-use implementation of the preference dialog handling alone.

**Comment by DreaminginCodeZH** [comment]
Yes, I agree the approach you proposed is totally viable and do has its advantages, but still I prefer a more flexible `addDialogPreference(Preference, Fragment)` (or `registerPreferenceDialogFragment` semantically speaking). And since one can always show a dialog in a plain `Fragment` (e.g. A regular fragment that does something and wraps a dialog or a dialog fragment in it for some reason), requiring the fragment used in this mechanism to be a subclass of `DialogFragment` is kind of rigid.

So maybe can you just make it happen? (Accepting the any Preference and Fragment, being deprecated is ok)

**Comment by Gericop** [comment]
@DreaminginCodeZH 
Released v**27.0.0.0**.
Use `PreferenceFragmentCompat.registerPreferenceFragment(...)` to register a preference of which the fragment should be added to the fragment manager when the preference is clicked.

**Comment by DreaminginCodeZH** [comment]
Thanks!

**Comment by DreaminginCodeZH** [comment]
Sorry to bother again, but can you call `setTargetFragment()` for a `Fragment` class just as what happens for a `DialogFragment` class? It is crucial in getting the `PreferenceFragment` and then the `Preference`, and this maintains the same behavior as `DialogFragment` without breaking anything.

And since I'm bothering again, the naming of `registerPreferenceFragment()` did not reflect the fact that it is for `showDialog()` and `onDisplayPreferenceDialog()`. Maybe `registerPreferenceDialogFragment()` is a better name?

**Comment by Gericop** [comment]
Yeah, `setTargetFragment(...)` can be added.

I think `registerPreferenceDialogFragment(...)` would be a good name if the second parameter was a `DialogFragment` class (which is not the case anymore). Maybe a `registerPreferenceFragmentForDialog(...)` or something, but I don't think the name matters too much as it's rarely used for custom preferences, most devs will use only the built-in ones to provide a settings screen quickly and easily. Also, the javadoc speaks for itself.
